pkg/mutate/squashfs.go

```diff
@@ -124,6 +141,53 @@ func (c *squashfsConverter) makeSquashfs(r io.Reader) (string, error) {
 	return path, nil
 }
 
+// convertWhiteout accepts a Layer l and returns:
+//
+//   - A ReadCloser providing the layer tar, passed through a filter that converts
+//     whiteout markers from AUFS -> OverlayFS, if c.noConvertWhiteout is false.
+//     In this case, any error from the filter will propagate via the returned channel.
+//   - A ReadCloser providing the layer tar directly, if c.noConvertWhiteout is true.
+//
+// Note that when conversion is performed, the layer is read twice.
+func (c *squashfsConverter) convertWhiteout(l v1.Layer) (io.ReadCloser, chan error, error) {
+	rc, err := l.Uncompressed()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// No conversion - direct tar stream from the layer.
+	if c.noConvertWhiteout {
+		return rc, nil, nil
+	}
+
+	// Conversion - first, scan for opaque directories and presence of file
+	// whiteout markers.
+	opaquePaths, fileWhiteout, err := scanAUFSWhiteouts(rc)
+	if err != nil {
+		return nil, nil, err
+	}
+	rc.Close()
+
+	rc, err = l.Uncompressed()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Nothing found to filter
+	if len(opaquePaths) == 0 && !fileWhiteout {
+		return rc, nil, nil
+	}
+
+	filterErr := make(chan error, 1)
+	pr, pw := io.Pipe()
+	go func() {
+		defer rc.Close()
+		filterErr <- whiteoutFilter(rc, pw, opaquePaths)
+		close(filterErr)
+	}()
+	return pr, filterErr, nil
+}
```

Comment:
Probably simpler to return the `error` from `whiteoutFilter` via the `io.Writer` using [`(*PipeWriter).CloseWithError`](https://pkg.go.dev/io#PipeWriter.CloseWithError)?

```suggestion

	pr, pw := io.Pipe()
	go func() {
		defer rc.Close()

		pw.CloseWithError(whiteoutFilter(rc, pw, opaquePaths))
	}()
	return pr, nil
```

Otherwise, the `io.ReadCloser` will return `EOF`, causing the caller to make the SquashFS only to find out that the conversion failed.

Reply:
Cheers - hadn't come across that before.